### PR TITLE
Implement filter creation functionality based on the map

### PR DESCRIPTION
**Pull request by jamal-khey**

### CODE REVIEW

File: src/translations/fr.json [hunk 1]
```diff
@@ -205,6 +205,8 @@
     "SelectAll": "Tous",
     "UnselectAll": "Aucun",
 
+    "DrawingPolygonInstruction": "Cliquez sur la carte pour commencer a dessiner un polygone. Entrer pour terminer ou echap pour annuler.",
```
Comment (sBouzols) [inline src/translations/fr.json:208]:
```suggestion
    "DrawingPolygonInstruction": "Cliquez sur la carte pour commencer à dessiner un polygone. Entrée pour terminer ou Échap pour annuler.",
```

File: src/translations/fr.json [hunk 2]
```diff
@@ -982,6 +986,7 @@
     "EquipmentCopied": "Données copiées depuis l'ouvrage {equipmentId}",
     "EquipmentCopyFailed404": "Erreur lors de la copie de l'ouvrage : l'ouvrage n'a pas été trouvé",
     "EquipmentCopyFailed": "Erreur lors de la copie de l'ouvrage",
+    "equipmentType": "Type d'ouvrage",
```
Comment (sBouzols) [inline src/translations/fr.json:989]: the is already an `EquipmentType` translation key for this.

File: src/translations/en.json [hunk 3]
```diff
@@ -981,6 +986,7 @@
     "EquipmentCopied": "Data successfully copied from equipment {equipmentId}",
     "EquipmentCopyFailed404": "Equipment copy failed: the equipment was not found",
     "EquipmentCopyFailed": "Equipment copy failed",
+    "equipmentType": "Equipment type",
```
Comment (sBouzols) [inline src/translations/en.json:989]: there is already an EquipmentType translation key for this.

File: src/services/study/network.js [hunk 4]
```diff
@@ -132,6 +132,42 @@ export function getSubstationSingleLineDiagram(
     );
 }
 
+export function fetchNetworkElementsIds(
```
Comment (sBouzols) [inline src/services/study/network.js:135]: Why not using the existing `fetchEquipmentsIds` in src/services/study/network-map.js

File: src/services/explore.ts [hunk 5]
```diff
@@ -113,3 +113,44 @@ export function createModifications(
...
+//         "equipmentID": "fds"
+//       },
+//       {
+//         "equipmentID": "sf"
+//       },
+//       {
+//         "equipmentID": "fes"
+//       }
+//     ]
+//   }
```
Comment (sBouzols) [inline src/services/explore.ts:135]: You will not maintain this comment if DTO change, it's not useful I think

File: src/redux/reducer.type.ts [hunk 6]
```diff
@@ -28,6 +28,8 @@ export interface ReduxState {
     notificationIdList: UUID[];
     theme: string;
     nonEvacuatedEnergyNotif: boolean;
+    geoData: any;
```
Comment (sBouzols) [inline src/redux/reducer.type.ts:31]: We should definitly define those types

File: src/redux/reducer.js [hunk 7]
```diff
@@ -284,6 +284,7 @@ const initialState = {
     ...paramsInitialState,
     limitReductionModified: false,
     lastCompletedComputation: null,
+    polygonCoordinate: [],
```
Comment (sBouzols) [inline src/redux/reducer.js:287]: Why do we need this in the store ?

File: src/components/map-view.jsx [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                                </Box>
+                            </Box>
+
+                            <DiagramPane
+                                studyUuid={studyUuid}
+                                showInSpreadsheet={showInSpreadsheet}
+                                currentNode={currentNode}
+                                visible={
+                                    view === StudyView.MAP &&
+                                    studyDisplayMode !== STUDY_DISPLAY_MODE.TREE
```
Comment (sBouzols) [inline src/components/map-view.jsx:236]: AND !== STUDY_DISPLAY_MODE.DRAW ?

Comment (sBouzols) [reply src/components/map-view.jsx:236]: see one usage in src/components/network-modification-tree-pane.jsx as well

File: src/components/study-pane.jsx [hunk 9]
```diff
@@ -260,7 +96,18 @@ const StudyPane = ({ studyUuid, currentNode, setErrorMessage, ...props }) => {
                     display: props.view === StudyView.MAP ? null : 'none',
                 }}
             >
-                {renderMapView()}
+                <MapView
```
Comment (sBouzols) [inline src/components/study-pane.jsx:99]: Even if it's obviously a good idea to create this component and move it in a proper file, it makes the review more difficult ! It's better to do a separate PR for those refactoring achievements.

File: src/components/map-view.jsx [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                                            lineFlowAlertThreshold
+                                        }
+                                        openVoltageLevel={openVoltageLevel}
+                                        /* TODO verif tableEquipment*/
+                                        currentNode={currentNode}
+                                        onChangeTab={onChangeTab}
+                                        showInSpreadsheet={showInSpreadsheet}
+                                        setErrorMessage={setErrorMessage}
+                                        onDrawModeChanged={(evt) => {
+                                            if (evt === true) {
```
[the remark this answers is not in the document]
Comment (sBouzols) [reply src/components/map-view.jsx:217]: or rename parameter to `active` and keep boolean type

File: src/redux/reducer.type.ts (hunk 6 above)
Comment (jamal-khey) [reply src/redux/reducer.type.ts:31]: i removed them for now,  we need to make @types/powsybl-diagram-viewer to be able to import them in reducer.type.ts

File: src/services/study/network.js (hunk 4 above)
Comment (jamal-khey) [reply src/services/study/network.js]: done , but i discovred a bug in 'fetchEquipmentsIds in src/services/study/network-map.js so i did some changes to it

File: src/redux/reducer.js (hunk 7 above)
Comment (jamal-khey) [reply src/redux/reducer.js]: removed

File: src/services/explore.ts (hunk 5 above)
Comment (jamal-khey) [reply src/services/explore.ts]: done

File: src/translations/en.json [hunk 11]
```diff
@@ -205,6 +205,8 @@
     "SelectAll": "All",
     "UnselectAll": "None",
 
+    "DrawingPolygonInstruction": "Click the map to draw a polygon. press enter to validate or escape to cancel",
```
[the remark this answers is not in the document]
Comment (jamal-khey) [reply src/translations/en.json]: done

File: src/translations/en.json (hunk 3 above)
Comment (jamal-khey) [reply src/translations/en.json]: done

File: src/translations/fr.json (hunk 1 above)
Comment (jamal-khey) [reply src/translations/fr.json]: done

File: src/services/study/network.js (hunk 4 above)
Comment (sBouzols) [reply src/services/study/network.js]: Keep second implementation and plan a refactor of the previous function to avoid impact on other functionalities.

File: src/components/map-view.jsx (hunk 8 above)
Comment (sBouzols) [reply src/components/map-view.jsx:236]: It should be ok.

File: src/components/study-pane.jsx (hunk 9 above)
Comment (sBouzols) [reply src/components/study-pane.jsx:99]: Ok keep this refacto in this PR

File: src/services/study/network-map.js [hunk 12]
```diff
@@ -137,3 +146,102 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+    studyUuid,
+    currentNodeUuid,
+    networkMapref
+) {
+    let equipementList = [];
+    switch (filter.equipmentType) {
+        case EQUIPMENT_TYPES.SUBSTATION:
+            equipementList = createEquipmentIdentifierList(
+                filter.equipmentType,
+                networkMapref.current.getSelectedSubstation()
```
Comment (sBouzols) [inline src/services/study/network-map.js:175]: Keep networkMapref near the component and pass selectedSubstation array to this function.

File: src/components/map-view.jsx [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+                                    studyDisplayMode === STUDY_DISPLAY_MODE.TREE
+                                        ? 'none'
+                                        : null,
+                                width: getMapWitdh(studyDisplayMode),
+                            }}
+                        >
+                            <Box>
+                                <Box sx={styles.mapBelowDiagrams}>
+                                    <NetworkMapTab
+                                        networkMapRef={networkMapref}
```
Comment (sBouzols) [inline src/components/map-view.jsx:195]: move the ref creation inside NetworkMapTab and add a prop to NetworkMapTab to reset the draw mode for example which will call networkMapref.current.getMapDrawer().deleteAll();

File: src/services/study/network-map.js [hunk 14]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+        default:
+            break;
+    }
+
+    if (equipementList.filterEquipmentsAttributes?.length === 0) {
+        throw new Error('No equipment selected');
+    }
+
+    if (equipementList.filterEquipmentsAttributes.length > 0) {
+        await createFilter(
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:231]: Deal with errors ?

File: src/services/study/network-map.js [hunk 15]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+            }
+        }),
+    };
+}
+export async function createMapFilter(
+    filter,
+    distDir,
+    studyUuid,
+    currentNodeUuid,
+    selectedEquipements
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:168]: Rzname by `selectedEquipementsIds`

File: src/services/study/network-map.js [hunk 16]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+        case EQUIPMENT_TYPES.LCC_CONVERTER_STATION:
+        case EQUIPMENT_TYPES.SWITCH:
+            const substationsIds = selectedEquipements.map(
+                (substation) => substation.id
+            );
+            if (substationsIds.length === 0) {
+                throw new Error('No substations selected');
+            }
+            try {
+                const elements = await fetchEquipmentsIds(
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:202]: Rename by `elementsIds`

File: src/services/study/network-map.js [hunk 17]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+    };
+}
+export async function createMapFilter(
+    filter,
+    distDir,
+    studyUuid,
+    currentNodeUuid,
+    selectedEquipements
+) {
+    let equipementList = [];
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:170]: Rename by `equipementFilters` ?

File: src/services/study/network-map.js [hunk 18]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+                );
+
+                if (elements.length > 0) {
+                    equipementList = createEquipmentIdentifierList(
+                        filter.equipmentType,
+                        elements
+                    );
+                }
+            } catch (error) {
+                throw error;
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:217]: ?

File: src/services/study/network-map.js [hunk 19]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
     console.debug(fetchCountriesUrl);
     return backendFetchJson(fetchCountriesUrl);
 }
+
+function createEquipmentIdentifierList(equipmentType, equipmentList) {
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:150]: Rename by `createEquipmentFilters`

Comment (SlimaneAmar) [inline src/services/study/network-map.js:150]: Rename by `equipmentIds` ?

File: src/services/study/network-map.js [hunk 20]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+    return {
+        type: 'IDENTIFIER_LIST',
+        equipmentType: equipmentType,
+        filterEquipmentsAttributes: equipmentList.map((eqId) => {
+            if (eqId?.id) {
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:155]: Why this test ?

File: src/services/study/network-map.js [hunk 21]
```diff
@@ -137,3 +146,93 @@ export function fetchAllCountries(studyUuid, currentNodeUuid) {
...
+        case EQUIPMENT_TYPES.BATTERY:
+        case EQUIPMENT_TYPES.LOAD:
+        case EQUIPMENT_TYPES.SHUNT_COMPENSATOR:
+        case EQUIPMENT_TYPES.DANGLING_LINE:
+        case EQUIPMENT_TYPES.STATIC_VAR_COMPENSATOR:
+        case EQUIPMENT_TYPES.HVDC_CONVERTER_STATION:
+        case EQUIPMENT_TYPES.VSC_CONVERTER_STATION:
+        case EQUIPMENT_TYPES.LCC_CONVERTER_STATION:
+        case EQUIPMENT_TYPES.SWITCH:
+            const substationsIds = selectedEquipements.map(
```
Comment (SlimaneAmar) [inline src/services/study/network-map.js:195]: Too many tests in theses functions
Pb design ?

File: src/components/network/filter-creation-panel.tsx [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+import {
+    fetchDirectoryContent,
+    fetchPath,
+    fetchRootFolders,
+} from 'services/directory';
+import { fetchElementsMetadata } from 'services/explore';
+import { UniqueNameInput } from 'components/dialogs/commons/unique-name-input';
+import { useSelector } from 'react-redux';
+
+const EXPERT_FILTER_EQUIPMENTS = {
```
[the remark this answers is not in the document]
Comment (jamal-khey) [reply src/components/network/filter-creation-panel.tsx:25]: we need the labels for traductions,  i moved this object to `equipment-type.js` close to EQUIPMENT_TYPES

File: src/services/study/network-map.js (hunk 14 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:231]: i deal with the errors in the caller 'map-viwer.jsx'

File: src/services/study/network-map.js (hunk 18 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:217]: removed

File: src/services/study/network-map.js (hunk 17 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:170]: done

File: src/services/study/network-map.js (hunk 16 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:202]: done

File: src/services/study/network-map.js (hunk 15 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:168]: done

File: src/components/network/filter-creation-panel.tsx [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+                            disableClearable={true}
+                            formProps={{ style: { fontStyle: 'italic' } }}
+                        />
+                    </Grid>
+
+                    <Grid container paddingTop={2}>
+                        <UniqueNameInput
+                            name={NAME}
+                            label={'Name'}
+                            elementType={'DIRECTORY'}
```
[the remark this answers is not in the document]
Comment (jamal-khey) [reply src/components/network/filter-creation-panel.tsx:201]: done

File: src/components/network/filter-creation-panel.tsx [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+    onSaveFilter: (data: IFilterCreation, distDir: Identifier) => void;
+    onCancel: () => void;
+};
+
+const FilterCreationPanel: React.FC<FilterCreationPanelProps> = ({
+    onSaveFilter,
+    onCancel,
+}) => {
+    const studyUuid = useSelector((state: any) => state.studyUuid);
+    const [openDirectoryFolders, setOpenDirectoryFolders] = useState(false);
```
[the remark this answers is not in the document]
Comment (jamal-khey) [reply src/components/network/filter-creation-panel.tsx:122]: done

File: src/services/study/network-map.js (hunk 19 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:147]: done

File: src/services/study/network-map.js (hunk 20 above)
Comment (jamal-khey) [reply src/services/study/network-map.js:152]: The list of equipment. Each equipment can be either a string representing the equipment ID or an object with an 'id' property.

i added  comments to explay the arguments

File: src/services/study/network-map.js (hunk 21 above)
Comment (jamal-khey) [reply src/services/study/network-map.js]: i improved a bit the code. with the usage of 'default'

i added the tests with diffirents errors so the user can have an explicit error message why the selection fails.
for the others test mostly .length > 0.  is due to the backend that have special case for when we pass empty array

File: src/translations/fr.json (hunk 1 above)
Comment (jamal-khey) [reply src/translations/fr.json]: done

**Review by SlimaneAmar — APPROVED** [review]
